### PR TITLE
[ObsTransforms] Add support for semantic sensor observation transforms

### PR DESCRIPTION
**Pull request by Ram81**
## Motivation and Context

Add support for semantic sensor in observation transformer to support ObjectNav DD-PPO training with semantic sensor.

## How Has This Been Tested

Locally by using DD-PPO training for ObjectNav

## Types of changes

- [x] Changes semantic observation dimensions to (H,W, 1) in `HabitatSimSemanticSensor` sensor
- [x] Add support for semantic sensor in `ResizeShortestEdge` observation transforms
- [x] Handle interpolation for semantic observation in `image_resize_shortest_edge` util method

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: habitat_baselines/common/obs_transformers.py [hunk 1]
```diff
@@ -84,6 +84,11 @@ def __init__(
         self._size: int = size
         self.channels_last: bool = channels_last
         self.trans_keys: Tuple[str, ...] = trans_keys
+        self.interpolation_mode = {
+            "rgb": "area",
+            "depth": "area",
+            "semantic": "nearest-exact",
```
Comment (Skylion007) [inline habitat_baselines/common/obs_transformers.py:90]: You may want to encode a default of area.

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 2]
```diff
@@ -191,15 +191,22 @@ def _get_observation_space(self, *args: Any, **kwargs: Any):
...
+            shape=(self.config.HEIGHT, self.config.WIDTH, 1),
+            dtype=np.int32,
         )
 
     def get_observation(
         self, sim_obs: Dict[str, Union[np.ndarray, bool, "Tensor"]]
     ) -> VisualObservation:
         obs = cast(Optional[VisualObservation], sim_obs.get(self.uuid, None))
         check_sim_obs(obs, self)
+        # make semantic observation a 3D array
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:203]:
```sugg

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 3]
```diff
@@ -191,15 +191,22 @@ def _get_observation_space(self, *args: Any, **kwargs: Any):
...
+        if isinstance(obs, np.ndarray):
+            obs = np.expand_dims(
+                obs, axis=2
+            ).astype(np.int32)
+        else:
+            obs = obs.unsqueeze(-1)
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:209]:
```suggestion
        obs = obs[..., None]
```

File: habitat_baselines/common/obs_transformers.py (hunk 1 above)
Comment (Ram81) [reply habitat_baselines/common/obs_transformers.py:90]: @Skylion007 but is it right to use the area for semantic sensor observations? It might create non-int values at some places which might not work if you are trying to pass semantic observations through a embedding layer in your policy

Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:90]: I don't like using a dict here because it's not robust to multiple RGB sensors or depth cameras etc which is a common use case.

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 4]
```diff
@@ -200,13 +200,7 @@ def get_observation(
...
         obs = cast(Optional[VisualObservation], sim_obs.get(self.uuid, None))
         check_sim_obs(obs, self)
-        # make semantic observation a 3D array
-        if isinstance(obs, np.ndarray):
-            obs = np.expand_dims(
-                obs, axis=2
-            ).astype(np.int32)
-        else:
-            obs = obs.unsqueeze(-1)
+        obs = obs[..., None].astype(np.int32)
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:203]: This won't work if it's a Tensor though, right?

**Review by Skylion007 — CHANGES_REQUESTED** [review]

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 4 above)
Comment (Ram81) [reply habitat/sims/habitat_simulator/habitat_simulator.py:203]: Yes, this won't work. Will there be a case when simulator returns a tensor object? In my experiments it has always been a numpy array. We need to typecast to int32 as tensor won't support uint32

File: habitat_baselines/common/obs_transformers.py (hunk 1 above)
Comment (Ram81) [reply habitat_baselines/common/obs_transformers.py:90]: @vincentpierre @Skylion007 if we want to move interpolation mode to config where do you recommend we put it? I am not sure how we can move the interpolation mode to `ENABLED_TRANSFORMS.RESIZE_SHORTEST_EDGE`, this config is specific to a sensor so we would need to put it for each sensor in the transforms config.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 4 above)
Comment (Skylion007) [reply habitat/sims/habitat_simulator/habitat_simulator.py]: When it's running in GPU2GPU mode.

File: habitat_baselines/common/obs_transformers.py (hunk 1 above)
Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:90]: Yeah, if I redesigned the OBS_TRANSFORMER I would have attached the keys to the sensors. You can specify the keys for individual transforms, but not individual args. @vincentpierre This probably would be a good aspect to redesign as it's a recurring problem.

@Ram81 We ran into this issue with Equirect ObsTransformers and used DTYPE as hack there to detect if it's was a semantic sensor I think

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 5]
```diff
@@ -191,15 +191,20 @@ def _get_observation_space(self, *args: Any, **kwargs: Any):
...
         )
 
     def get_observation(
         self, sim_obs: Dict[str, Union[np.ndarray, bool, "Tensor"]]
     ) -> VisualObservation:
         obs = cast(Optional[VisualObservation], sim_obs.get(self.uuid, None))
         check_sim_obs(obs, self)
+        # make semantic observation a 3D array
+        if isinstance(obs, np.ndarray):
+            obs = obs[..., None].astype(np.int32)
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:205]: Why do we convert the type to int32 here? Confused why it's necessary.

File: habitat_baselines/common/obs_transformers.py (hunk 1 above)
Comment (vincentpierre) [reply habitat_baselines/common/obs_transformers.py:90]: I thought there was a clean way to do this. If there is none, maybe what you have is the best solution

Comment (Skylion007) [reply habitat_baselines/common/obs_transformers.py:90]: The "clean" way of doing it previously was detecting dtype, but we really should have a cleaner way to do it.

Comment (Ram81) [reply habitat_baselines/common/obs_transformers.py:90]: @vincentpierre @Skylion007 yes, currently. This looked like the only way to do this. I can look into how `Equirect ObsTransformers` handles this and see if I can make it any cleaner than this.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 5 above)
Comment (Ram81) [reply habitat/sims/habitat_simulator/habitat_simulator.py:205]: @Skylion007 if I don't convert it here then I'll have to do it in the trainer. Torch doesn't support uint32 so when you initialize the rollout storage with semantic observations that have type np.uint32 it breaks. I moved the conversion here to avoid doing it every time in trainer.

Comment (Skylion007) [reply habitat/sims/habitat_simulator/habitat_simulator.py:205]: Interesting, is this a bug in our habitat-sim code? WE should be returning the same dtype for numpy as torch if possible?

Comment (Ram81) [reply habitat/sims/habitat_simulator/habitat_simulator.py:205]: @Skylion007 I think so. In my experiments I had to manually typecast to np.int32 to use semantic input for the policy

**Comment by Ram81** [comment]
@Skylion007 the mypy check is failing on [this line](https://github.com/facebookresearch/habitat-lab/pull/847/files#diff-f71c4b18459948bb12a64dbd8cb0f2e49530ae119e83cdd2a07605690536feaeR207). Does this mean semantic sensor is not supported in GPU2GPU mode? So we don't have to handle the case when it's a tensor?

**Comment by Ram81** [comment]
Also, I am not sure why black is failing with pre-commit. Is there because of dependency conflict issue?

**Review by vincentpierre — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -48,6 +48,7 @@ repos:
     hooks:
     - id: black
       exclude: ^examples/tutorials/(nb_python|colabs)
+      additional_dependencies: ['click==8.0.4']
```
Comment (Skylion007) [inline .pre-commit-config.yaml:51]:
```suggestion
```

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 7]
```diff
@@ -191,15 +191,18 @@ def _get_observation_space(self, *args: Any, **kwargs: Any):
...
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:205]:
```suggestion
            obs = obs[..., None].astype(np.int32)
    else:
            obs = obs[..., None]
```
PyTorch tensors can still be returned as observations if the GPU2GPU flag is enabled

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Hmm, so apparently our test suite actually does not test the edge case of using a semantic sensor with gpu2gpu enabled for a ResizeCrop/CenterCrop (in test_baseline_trainers). That's why this bug didn't get caught, we may want to rectify that issue.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 7 above)
Comment (Ram81) [reply habitat/sims/habitat_simulator/habitat_simulator.py:205]: @Skylion007 the mypy tests fail if I handle torch tensor. It says the statement is unreachable that's why I removed change on L207

Comment (Skylion007) [reply habitat/sims/habitat_simulator/habitat_simulator.py:205]: @Ram81 An easy way to get this merged is to do obs = obs[..., None] before the isinstance and just do the as type after

**Review by Skylion007 — APPROVED** [review]
